compositor_render/src/transformations/shader/validation.rs

```diff
@@ -0,0 +1,418 @@
+    VertexShaderBadInputTypeName(Option<String>),
+
+    #[error("the vertex shader input has a wrong type")]
+    VertexShaderBadInput(#[source] TypeEquivalenceError),
+}
+
+pub fn validate_contains_header(
+    header: &naga::Module,
+    shader: &naga::Module,
+) -> Result<(), ShaderValidationError> {
+    validate_globals(header, shader)?;
+
+    validate_vertex_input(header, shader)?;
+
+    Ok(())
+}
+
+fn validate_globals(
+    header: &naga::Module,
+    shader: &naga::Module,
+) -> Result<(), ShaderValidationError> {
+    for (_, global) in header.global_variables.iter() {
+        let Some((_, global_in_shader)) = shader.global_variables.iter().find(|(_, s_global)| {
+            s_global.space == global.space && s_global.binding == global.binding
+        }) else {
+            return Err(ShaderValidationError::GlobalNotFound(global.clone()));
+        };
+
+        type_equivalent(global.ty, header, global_in_shader.ty, shader)
+            .map_err(ShaderValidationError::GlobalBadType)?;
+    }
+
+    Ok(())
+}
+
+fn validate_vertex_input(
+    header: &naga::Module,
+    shader: &naga::Module,
+) -> Result<(), ShaderValidationError> {
+    let vertex = shader
+        .entry_points
+        .iter()
+        .find(|entry_point| {
+            entry_point.name == super::VERTEX_ENTRYPOINT_NAME
+                && entry_point.stage == ShaderStage::Vertex
+        })
+        .ok_or(ShaderValidationError::VertexShaderNotFound)?;
+
+    if vertex.function.arguments.len() != 1 {
+        return Err(ShaderValidationError::VertexShaderBadArgumentAmount(
+            vertex.function.arguments.len(),
+        ));
+    }
+
+    let vertex_input = vertex.function.arguments[0].ty;
+    let vertex_input_type = &shader.types[vertex_input];
+
+    let (header_vertex_input, _) = header
+        .types
+        .iter()
+        .find(|(_, ty)| ty.name == vertex_input_type.name)
+        .ok_or(ShaderValidationError::VertexShaderBadInputTypeName(
+            vertex_input_type.name.clone(),
+        ))?;
+
+    type_equivalent(header_vertex_input, header, vertex_input, shader)
+        .map_err(ShaderValidationError::VertexShaderBadInput)?;
+
+    Ok(())
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum TypeEquivalenceError {
+    #[error("type names don't match: {0:?} != {1:?}")]
+    TypeNameMismatch(Option<String>, Option<String>),
+
+    #[error("type internal structure doesn't match: {0:?} != {1:?}")]
+    TypeStructureMismatch(naga::TypeInner, naga::TypeInner),
+
+    #[error("sizes of an array don't match: {0:?} != {1:?}")]
+    ArraySizeMismatch(ArraySizeOrConstant, ArraySizeOrConstant),
+
+    #[error("a composite type was used as an array length specifier")]
+    // don't think this will ever happen
+    CompositeTypeAsArrayLen(ConstantInner),
+}
+
+#[derive(Debug)]
+pub enum ArraySizeOrConstant {
+    ArraySize(ArraySize),
+    Constant(Constant),
+}
+
+fn type_equivalent(
```

Comment:
check_type_equivalent
or
validate_type_equivalent
or 
is_type_equivalen (and return type to bool)